File: dungeons-26.1.2/src/main/java/fishmod/features/PbPaceHud.java
package fishmod.features;

import fishmod.utils.Constants;
import fishmod.utils.config.values.FishSettings;
import fishmod.utils.dungeon.Phase;
import fishmod.utils.dungeon.RunHistory;
import fishmod.utils.dungeon.Split;
import net.minecraft.client.DeltaTracker;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.screens.ChatScreen;

/**
 * PB Pace — a racing-style "ghost" for dungeon runs. As each split completes it compares your live
 * time to your personal best for that split and shows a running delta: green when you're ahead of
 * your PB pace, red when you're behind. Pure read-over of the existing split + run-history systems,
 * so it costs nothing until you actually start a run with recorded history.
 */
public final class PbPaceHud {

    private PbPaceHud() {}

    public static boolean isVisible() {
        return FishSettings.pbPaceEnabled && Phase.runStarted() && !Phase.getCurrentSplits().isEmpty();
    }

    public static void renderHud(GuiGraphicsExtractor ctx, DeltaTracker tick) {
        if (!FishSettings.pbPaceEnabled) return;
        Minecraft mc = Minecraft.getInstance();
        if (mc.player == null) return;
        if (mc.screen != null && !(mc.screen instanceof ChatScreen)) return;
        if (!Phase.runStarted()) return;

        java.util.List<Split> splits = Phase.getCurrentSplits();
        if (splits.isEmpty()) return;
        String floor = Phase.getFloor();

        double cumDelta = 0;
        boolean anyPb = false;
        Split last = null;
        double lastDelta = 0;
        boolean lastHasPb = false;

        for (Split s : splits) {
            if (!s.ended() || s.getAvg() < 0) continue; // skip unfinished + cumulative/total rows
            last = s;
            double pb = RunHistory.getPersonalBest(floor, s.getName());
            if (pb > 0) {
                double d = s.getRealTime() - pb;
                cumDelta += d;
                anyPb = true;
                lastDelta = d;
                lastHasPb = true;
            } else {
                lastHasPb = false;
            }
        }
        if (last == null) return; // no split finished yet — nothing to pace against

        int x = FishSettings.pbPaceHudX, y = FishSettings.pbPaceHudY;
        float sc = (float) FishSettings.pbPaceScale;
        int lh = Constants.TEXT_HEIGHT + 1;

        ctx.pose().pushMatrix();
        ctx.pose().translate((float) x, (float) y);
        ctx.pose().scale(sc, sc);
        int row = 0;
        ctx.text(mc.font, "§6§lPB Pace §7(" + floor.toUpperCase() + ")", 0, row++ * lh, 0xFFFFFFFF, true);
        if (lastHasPb)
            ctx.text(mc.font, "§f" + last.getName() + " " + signed(lastDelta), 0, row++ * lh, 0xFFFFFFFF, true);
        if (anyPb)
            ctx.text(mc.font, "§7vs PB: " + (cumDelta <= 0 ? "§aahead " : "§cbehind ") + signed(cumDelta), 0, row++ * lh, 0xFFFFFFFF, true);
        else
            ctx.text(mc.font, "§8building PB history…", 0, row++ * lh, 0xFFFFFFFF, true);
        ctx.pose().popMatrix();
    }

    /** Format a delta vs PB: green & "-" when faster, red & "+" when slower. */
    private static String signed(double d) {
        String num = Constants.DECIMAL_FORMAT.format(Math.abs(d));
        return d <= 0 ? "§a-" + num + "s" : "§c+" + num + "s";
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/features/PbPaceHud.kt
package fishmod.features

import fishmod.utils.Constants
import fishmod.utils.config.values.FishSettings
import fishmod.utils.dungeon.Phase
import fishmod.utils.dungeon.RunHistory
import fishmod.utils.dungeon.Split
import net.minecraft.client.DeltaTracker
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.screens.ChatScreen

/** Racing-style ghost: compares live split times to your personal best and shows a running delta. */
object PbPaceHud {

    @JvmStatic
    fun isVisible(): Boolean =
        FishSettings.pbPaceEnabled && Phase.runStarted() && Phase.getCurrentSplits().isNotEmpty()

    @JvmStatic
    fun renderHud(ctx: GuiGraphicsExtractor, tick: DeltaTracker) {
        if (!FishSettings.pbPaceEnabled) return
        val mc = Minecraft.getInstance()
        if (mc.player == null) return
        if (mc.screen != null && mc.screen !is ChatScreen) return
        if (!Phase.runStarted()) return

        val splits: List<Split> = Phase.getCurrentSplits()
        if (splits.isEmpty()) return
        val floor = Phase.getFloor()

        var cumDelta = 0.0
        var anyPb = false
        var last: Split? = null
        var lastDelta = 0.0
        var lastHasPb = false

        for (s in splits) {
            if (!s.ended() || s.avg < 0) continue // skip unfinished + cumulative/total rows
            last = s
            val pb = RunHistory.getPersonalBest(floor, s.name)
            if (pb > 0) {
                val d = s.getRealTime() - pb
                cumDelta += d
                anyPb = true
                lastDelta = d
                lastHasPb = true
            } else {
                lastHasPb = false
            }
        }
        if (last == null) return // no split finished yet — nothing to pace against

        val x = FishSettings.pbPaceHudX
        val y = FishSettings.pbPaceHudY
        val sc = FishSettings.pbPaceScale.toFloat()
        val lh = Constants.TEXT_HEIGHT + 1

        ctx.pose().pushMatrix()
        ctx.pose().translate(x.toFloat(), y.toFloat())
        ctx.pose().scale(sc, sc)
        var row = 0
        ctx.text(mc.font, "§6§lPB Pace §7(" + (floor?.uppercase() ?: "") + ")", 0, row++ * lh, 0xFFFFFFFF.toInt(), true)
        if (lastHasPb)
            ctx.text(mc.font, "§f" + last.name + " " + signed(lastDelta), 0, row++ * lh, 0xFFFFFFFF.toInt(), true)
        if (anyPb)
            ctx.text(mc.font, "§7vs PB: " + (if (cumDelta <= 0) "§aahead " else "§cbehind ") + signed(cumDelta), 0, row++ * lh, 0xFFFFFFFF.toInt(), true)
        else
            ctx.text(mc.font, "§8building PB history…", 0, row++ * lh, 0xFFFFFFFF.toInt(), true)
        ctx.pose().popMatrix()
    }

    private fun signed(d: Double): String {
        val num = Constants.DECIMAL_FORMAT.format(Math.abs(d))
        return if (d <= 0) "§a-" + num + "s" else "§c+" + num + "s"
    }
}
